arroyo/processing/strategies/routing_producer.py

```diff
@@ -0,0 +1,155 @@
+    """
+
+    def __init__(
+        self,
+        commit_function: Commit,
+        commit_max_batch_size: int,
+        commit_max_batch_time: float,
+        message_router: MessageRouter,
+    ) -> None:
+        self.__commit_function = commit_function
+        self.__commit_max_batch_size = commit_max_batch_size
+        self.__commit_max_batch_time = commit_max_batch_time
+        self.__message_router = message_router
+        self.__all_producers = self.__message_router.populate_routes()
+        self.__closed = False
+        self.__offsets_to_be_committed: MutableMapping[Partition, Position] = {}
+        self.__messages_processed_since_last_commit = 0
+        self.__started = time.time()
+        self.__queue: Deque[
+            Tuple[Message[KafkaPayload], Future[Message[KafkaPayload]]]
+        ] = deque()
+
+    def _should_commit(self) -> bool:
+        now = time.time()
+        duration = now - self.__started
+        if self.__messages_processed_since_last_commit >= self.__commit_max_batch_size:
+            logger.info(
+                f"Max size reached: total of {self.__messages_processed_since_last_commit} messages after {duration:.{2}f} seconds"
+            )
+            return True
+        if now >= (self.__started + self.__commit_max_batch_time):
+            logger.info(
+                f"Max time reached: total of {self.__messages_processed_since_last_commit} messages after {duration:.{2}f} seconds"
+            )
+            return True
+
+        return False
+
+    def poll(self) -> None:
+        """
+        KafkaProducer's call poll() periodically in a thread. So we don't
+        need to explicitly call poll() here.
+        """
+        while self.__queue:
+            message, future = self.__queue[0]
+
+            if not future.done():
+                break
+
+            exc = future.exception()
+            if exc is not None:
+                raise exc
+
+            self.__queue.popleft()
+            self.__messages_processed_since_last_commit += 1
+            self.__offsets_to_be_committed[message.partition] = Position(
+                message.next_offset, message.timestamp
+            )
+
+        if self._should_commit():
+            self.__commit_function(self.__offsets_to_be_committed)
+            self.__messages_processed_since_last_commit = 0
+            self.__offsets_to_be_committed = {}
+            self.__started = time.time()
+
+    def submit(self, message: Message[KafkaPayload]) -> None:
+        assert not self.__closed
+        producer, topic = self.__message_router.get_route_for_message(message)
+        self.__queue.append(
+            (message, producer.produce(destination=topic, payload=message.payload))
+        )
+
+    def terminate(self) -> None:
+        self.__closed = True
+
+    def close(self) -> None:
+        self.__closed = True
+        for producer in self.__all_producers:
+            producer.close()
```

Comment:
Since the producer is created outside the strategy and passed in, I'm not sure it's the strategy's job to close it. It would be easy to accidentally try to produce again on a closed producer. I think I would prefer either:
- have the strategy also create the producer, so closing here it is always safe
- close the producer outside wherever it gets created